meteor/utils.py

```diff
@@ -75,6 +80,82 @@ def canonicalize_amplitudes(
         return None
 
 
+def rs_dataseies_to_complex_array(amplitudes: rs.DataSeries, phases: rs.DataSeries) -> np.ndarray:
+    """
+    Convert structure factors from polar (amplitude/phase) to Cartisian (x + iy).
+
+    Parameters
+    ----------
+    amplitudes: DataSeries
+        with StructureFactorAmplitudeDtype
+    phases: DataSeries
+        with PhaseDtype
+
+    Returns
+    -------
+    complex_structure_factors: np.ndarray
+        with dtype complex128
+
+    Raises
+    ------
+    ValueError
+        if the indices of `amplitudes` and `phases` do not match
+    """
+    try:
+        assert_index_equal(amplitudes.index, phases.index)
+    except AssertionError as exptn:
+        raise ShapeMismatchError(
+            f"indices for `amplitudes` and `phases` don't match: {amplitudes.index} {phases.index}",
+            " To safely cast to a single complex array, pass DataSeries with a common set of",
+            " indices. One possible way: Series.align(other, join='inner', axis=0).",
+            exptn,
+        )
+    complex_structure_factors = amplitudes.to_numpy() * np.exp(1j * np.deg2rad(phases.to_numpy()))
+    return complex_structure_factors
+
+
+def complex_array_to_rs_dataseries(
+    complex_structure_factors: np.ndarray,
+    index: pd.Index,
+) -> tuple[rs.DataSeries, rs.DataSeries]:
```

Comment:
Here we need to assume that the correct correspondence between hkl indices and structure factors was maintained?

Reply:
yep that's the point -- we loose `index` (HKL) info when moving to `np` arrays, so here you have to add it back again

It might be possible to work entirely with `DataSeries` of complex types that maintain the index, but many of `np`'s functions seem to have strange behavior when applied to `DataSeries`. Hence: dropping the index temporarily and adding it back later.